Template pdf

JosephBARBIERDARNAL opened the pull request.

JosephBARBIERDARNAL commented:
@dgkeyes Ok I think most of the things are implemented. See current template: [skeleton.pdf](https://github.com/user-attachments/files/20123073/skeleton.pdf)

TODO:

- [ ] cover page image on first page
- [ ] full width issue
- [ ] change symbol in bullet points

dgkeyes commented:
Looking good @JosephBARBIERDARNAL. Great work!

AlbertRapp commented:
Hi @JosephBARBIERDARNAL,

I took a look into this and indeed it was a tricky thing to work out why the things are not on full width. After a while I've realized that it didn't have anything to do with our CSS. Even if you were to eliminate all the CSS instructions from the CSS file, you'd still have a much larger margin on the right inside the PDF.

After some more playing around, I've realized that the `pagedown::html_paged()` function takes a `css` argument. If you insert the CSS file into that function and delete the `<link>` inside the body of the Rmd-file, then things work out and everything becomes full width. (All my changes were pushed to the branch.)

Unfortunately, this destroyed some styling wrt. to the order of the table of content and the abstract as well the styling of the bullet points. I didn't have time yet to figure out what's going on there but I suspect that using the `css` argument inside `pagedown::html_paged()` function is the intended way of using pagedown. Does that help you and can you look into why the bullet point styling was destroyed due to that? :thinking:

JosephBARBIERDARNAL commented:
![Uploading image.png…]()

JosephBARBIERDARNAL commented:
- toc titles: 01 instead of 1

JosephBARBIERDARNAL commented:
@AlbertRapp Thanks for the fix! Yes I also think it should use that `css` argument instead of what I was doing. The other things that broke are okay, I can fix it.

JosephBARBIERDARNAL commented:
> @JosephBARBIERDARNAL Is the cover page image issue still open? Just checking whether that's currently on my to-do list or yours :)

I've made some progress, but still not satisfying. Currently it covers well the page, but content below is hidden. I've tried to play with `z-index`, but then it looks weird. If you can have a look it's more than welcome :)

JosephBARBIERDARNAL commented:
> I've made some progress, but still not satisfying. Currently it covers well the page, but content below is hidden. I've tried to play with `z-index`, but then it looks weird. If you can have a look it's more than welcome :)

@AlbertRapp actually, when looking at the report on my computer (compared to google drive), it seems like it's supposed to hide elements like page number and logo at the bottom left, so I think it's okay. Let me get back to you on this.

JosephBARBIERDARNAL commented:
@AlbertRapp Some updates on the cover page(s):

There are 2 cases:

- cover page on the very first page, which is now working properly
- cover pages in the middle of the document, as stated in this issue https://github.com/rfortherestofus/omni/issues/121

For the second case, it should look like this:

![Screenshot 2025-05-14 at 17 06 41](https://github.com/user-attachments/assets/e7d5ba7b-1545-4a2a-8fa4-031f8b0015a3)

I started working on it, and at the moment, it looks like this (the background pattern is not the right one, but it's expected and easy to change):


<img width="534" alt="Screenshot 2025-05-14 at 17 07 13" src="https://github.com/user-attachments/assets/0ea53ac9-9e95-435e-8605-2927e07b7111" />

The thing I'm having trouble to do is to make the footer elements (page counter, bottom left logo, etc) to be visible. Also they should have different color style (mostly white). I'm not entirely what would be the best way to do this. Maybe it's just playing with `z-index` and some "color-hacking", but I haven't be able to do it yet.

If you can have a look it would be very helpful, thanks! Let me know if you have any question. (most relevant css is `.cover-page-full` and `.cover-content`)

@dgkeyes Do you if it's needed to be able to change the background pattern (choosing the red one, green one, etc)?

dgkeyes commented:
Yes, we do need to be able to change the background pattern.

dgkeyes commented:
@AlbertRapp hoping to get your help on this:

> The thing I'm having trouble to do is to make the footer elements (page counter, bottom left logo, etc) to be visible. Also they should have different color style (mostly white). I'm not entirely what would be the best way to do this. Maybe it's just playing with z-index and some "color-hacking", but I haven't be able to do it yet.

AlbertRapp commented:
Hi @dgkeyes and @JosephBARBIERDARNAL, sorry on the late reply. Had a few busy days. I looked into the issue of the footers not working properly. I've created a version that allows to display footers. The main fix that made that possible is to

- give `.pagedjs_pagebox .pagedjs_margin-bottom` a `z-index` of 2000 so that it is on top of the background image (which lives in a adjacent container)
- make all the margin containers use transparent background (rather than the default background color)

That way, the footer content can be displayed:

![image](https://github.com/user-attachments/assets/e3e86658-7b1f-4d5c-96c6-56d7da63b14e)

As for the styling of the footer, I managed to target the correct containers with JS and add the `custom-style` class that takes care of the styling. It looks like this:

```js
var pages = document.querySelectorAll('.cover-page-full')

pages.forEach(page => {
  var page_container = page.closest('.pagedjs_page');
  page_container.querySelector('.pagedjs_margin-bottom').classList.add('custom-style');
  page_container.querySelector('.pagedjs_margin.pagedjs_margin-bottom-center.hasContent').classList.add('custom-style');
})
```

Unfortunately, I haven't been able to execute it at the proper time when the page is fully rendered. This causes the css classes not to be attached. Still trying to figure out how the [Paged Handlers](https://pagedjs.org/documentation/10-handlers-hooks-and-custom-javascript/) work so that I can fix this. But for now this is still an open issue. Current preview is pushed to the `template-pdf` branch.

JosephBARBIERDARNAL commented:
> Unfortunately, I haven't been able to execute it at the proper time when the page is fully rendered. This causes the css classes not to be attached. Still trying to figure out how the [Paged Handlers](https://pagedjs.org/documentation/10-handlers-hooks-and-custom-javascript/) work so that I can fix this. But for now this is still an open issue. Current preview is pushed to the `template-pdf` branch.

I already added some custom JS and I had the same issue. It's in `fix-toc.html` and is added in the via `includes: in_header`:

```html
<script>
  setTimeout(() => {
    // Select all spans with class 'toc-section-number' inside the TOC
    const tocNumbers = document.querySelectorAll("#TOC .toc-section-number");

    tocNumbers.forEach((span) => {
      const num = parseInt(span.textContent.trim(), 10);
      if (!isNaN(num) && num >= 1 && num <= 9) {
        span.textContent = `0${num}`;
      }
    });
  }, 2000); // Adjust if rendering is delayed
</script>
```

EDIT: Ok I think I made your code to work. I made a few changes:

- I remove the custom style for the cover page (because we want to hide the footer in this case)
- I changed a bit your code and put it in `pdf_report_js.html` (I renamed `fix-toc.html` and combined our code) 
- It now looks like this:

<img width="1343" alt="Screenshot 2025-05-24 at 09 17 49" src="https://github.com/user-attachments/assets/2d915fde-9481-4fe9-96c7-ee68906aace7" />

AlbertRapp commented:
@JosephBARBIERDARNAL Nice! Does `setTimeout()` work reliably? :thinking:  I was thinking of using it but I wasn't sure whether it guarantees that the content is already rendered properly. Have you had trouble with this so far?

JosephBARBIERDARNAL commented:
> Nice! Does `setTimeout()` work reliably? 🤔 I was thinking of using it but I wasn't sure whether it guarantees that the content is already rendered properly. Have you had trouble with this so far?

Honestly I don't think it's the best way of handling it, but I haven't found a better way (and at the moment it has always work well for me when rendering it). Note that it's not needed for your code. 

Actually I'm thinking that maybe I should use the same logic as you (using pagedjs dedicated tools instead of some hacking like I did).

AlbertRapp commented:
> It now looks like this:

@JosephBARBIERDARNAL I cannot reproduce this. When I pull the latest version of this branch, the content in my footer doesn't turn white. Am I doing something wrong? I'm only rendering the `skeleton.Rmd` file.

dgkeyes commented:
Thanks @JosephBARBIERDARNAL and @AlbertRapp for your work on this. Can someone please summarize for me and @OskaratOmni what, if anything, remains to be done in this PR in order for us to merge it? I know there are issues related to the PDF template (e.g. #129 #130 and #132) that we will handle later, but please let us know if the v1 version of the template is ready.

JosephBARBIERDARNAL commented:
> Thanks @JosephBARBIERDARNAL and @AlbertRapp for your work on this. Can someone please summarize for me and @OskaratOmni what, if anything, remains to be done in this PR in order for us to merge it? I know there are issues related to the PDF template (e.g. #129 #130 and #132) that we will handle later, but please let us know if the v1 version of the template is ready.

I think we can merge.
